rfd/0095-teleport-key-agent-extension.md

```diff
@@ -0,0 +1,251 @@
+        Username              string `json:"user,omitempty"`
+        SiteName              string `json:"cluster,omitempty"`
+        ForwardedPorts        []string `json:"forward_ports,omitempty"`
+        DynamicForwardedPorts []string `json:"dynamic_forward_ports,omitempty"`
+        Dir                   string `json:"dir,omitempty"`
+        TLSRoutingEnabled     bool `json:"tls_routing_enabled,omitempty"`
+        AuthConnector         string `json:"auth_connector,omitempty"`
+        LoadAllCAs            bool `json:"load_all_cas,omitempty"`
+        MFAMode               string `json:"mfa_mode,omitempty"`
+}
+
+type ClientCertificates struct {
+        KeyIndex `json:"key_index"`
+        SSHCert      []byte `json:"ssh_cert"`
+        TLSCert      []byte `json:"tls_cert"`
+        TrustedCerts []TrustedCerts `json:"trusted_certs"`
+}
+
+type KeyIndex struct {
+        ProxyHost   string `json:"proxy_host"`
+        Username    string `json:"username"`
+        ClusterName string `json:"cluster_name"`
+}
+
+type TrustedCerts struct {
+        ClusterName      string `json:"domain_name"`
+        HostCertificates [][]byte `json:"checking_keys"`
+        TLSCertificates  [][]byte `json:"tls_certs"`
+}
+```
+
+Note: The underlying `teleport/api/profile.Profile` struct is subject to change, so there is a possibility of backwards compatibility concerns. However, this backwards compatibility should be handled already since the same concern is already applied for `~/.tsh` profile yaml files.
+
+**Sign Extension:**
+
+The standard `SSH_AGENTC_SIGN_REQUEST` expects to receive an un-hashed cryptograpic message challenge, and performs the hash and signature together on the ssh-agent server side. In `x/crypto`, this is sufficient for the `ssh.Signer` interface, but most uses of the `crypto.Signer` interface send hashed digests. This means that the `SSH_AGENTC_SIGN_REQUEST` can not be wrapped into a `crypto.Signer` without making modifications to the common `x/crypto` libraries. Additionally, some signature schemes are not supported by the `x/crypto` implementation of ssh-agent, including the `RSAPSS` signature schemes used in TLS 1.3.
+
+Instead, we will introduce a new sign extension that is more interoperable with the `crypto.Signer` interface.
+
+This extension requests a signature for the provided certs.
+
+        byte            SSH_AGENTC_EXTENSION
+        string          sign@goteleport.com
+        byte[]          key blob (ssh public key/certificate)
+        byte[]          digest
+        string          hash name
+        string          salt length
+
+The resulting signature will be returned alongside the signing algorithm used.
+
+        byte            SSH_AGENT_SIGN_RESPONSE
+        string          signature format
+        byte[]          signature blob
+
+Hash name should be the stringified representation of a golang `crypto.Hash` value. For example, `SHA-1`, `SHA-256`, or `SHA-512`.
```

Comment:
Do we have a complete list of supported hash functions? 
Regarding SHA-1, NIST formally deprecated the use of SHA-1 in 2011, so we should not use it in our examples.